enzyme/Enzyme/MLIR/Analysis/DataFlowActivityAnalysis.cpp

```diff
@@ -0,0 +1,976 @@
+//===- DataFlowActivityAnalysis.h - Implementation of Activity Analysis ---===//
+//
+//                             Enzyme Project
+//
+// Part of the Enzyme Project, under the Apache License v2.0 with LLVM
+// Exceptions. See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// If using this code in an academic setting, please cite the following:
+// @incollection{enzymeNeurips,
+// title = {Instead of Rewriting Foreign Code for Machine Learning,
+//          Automatically Synthesize Fast Gradients},
+// author = {Moses, William S. and Churavy, Valentin},
+// booktitle = {Advances in Neural Information Processing Systems 33},
+// year = {2020},
+// note = {To appear in},
+// }
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains the implementation of Activity Analysis -- an AD-specific
+// analysis that deduces if a given instruction or value can impact the
+// calculation of a derivative. This file formulates activity analysis within
+// a dataflow framework.
+//
+//===----------------------------------------------------------------------===//
+#include "DataFlowActivityAnalysis.h"
+#include "AliasAnalysis.h"
+#include "Dialect/Ops.h"
+#include "Interfaces/AutoDiffTypeInterface.h"
+
+#include "mlir/Analysis/DataFlow/ConstantPropagationAnalysis.h"
+#include "mlir/Analysis/DataFlow/DeadCodeAnalysis.h"
+#include "mlir/Analysis/DataFlow/DenseAnalysis.h"
+#include "mlir/Analysis/DataFlow/SparseAnalysis.h"
+#include "mlir/Analysis/DataFlowFramework.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+
+// TODO: Don't depend on specific dialects
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+
+#include "mlir/Analysis/AliasAnalysis/LocalAliasAnalysis.h"
+
+using namespace mlir;
+using namespace mlir::dataflow;
+using enzyme::AliasClassLattice;
+
+/// From LLVM Enzyme's activity analysis, there are four activity states.
+// constant instruction vs constant value, a value/instruction (one and the same
+// in LLVM) can be a constant instruction but active value, active instruction
+// but constant value, or active/constant both.
+
+// The result of activity states are potentially different for multiple
+// enzyme.autodiff calls.
+
+enum class ActivityKind { Constant, ActiveVal, Unknown };
```

Comment:
I'm not sure that there's a meaningful difference between an optimistic initialization to constant and explicitly processing a value as constant. I'm certainly not opposed to an `Undefined` state, I just don't immediately see the benefit.

Reply:
Thinking about the `setToEntryState` mess... Are mlir analysis supposed to be optimistic?

Reply:
Concretely, I know liveness analysis, which has an implementation upstream, [is initialized optimistically assuming all values are dead.](https://github.com/llvm/llvm-project/blob/4233176f9bf8d0309c03b127aa113ae826718dfa/mlir/include/mlir/Analysis/DataFlow/LivenessAnalysis.h#L66C3-L69C23)

Thinking about it more, I may have slightly misinterpreted the role of `setToEntryState`. It seems that its role is to *set to the pessimistic state* ($\top$), and the framework assumes that public function arguments should have this pessimistic state. This assumption isn't necessarily true for activity analysis because the activity of top-level function arguments should be explicitly provided by the user, and I believe most AD tools will take these user annotations as the truth even if dataflow can't be reasoned about at all.

Reply:
Basically, I have been trying (and so far mostly failing) to classify if the various analysis structures in terms of the following terminology. Optimistic analyses initialize lattice instances as the pessimistic fixpoint ($\top$) and transfer function may meet ($\cap$) lattice instances and progress towards optimistic fixpoint ($\bot$). Pessimistic analyses initialize lattice instances as the optimistic fixpoint ($\bot$) and progress towards $\top$ using join ($\cup$).

In the framework, forward analyses use join and backward analyses use meet, but I don't believe forward classes are meant for pessimistic and backward classes are meant for optimistic analyses. This may be a misinterpretation of moving "forward" (towards $\top$) or "backward" (towards $\bot$) along the lattice in the original implementation. I *think* forward and backward should only relate to the direction in which state is propagated, and both should be compatible with optimistic and pessimistic analyses. In fact, the `setToEntry/ExitState` should be more of a `setToFixPointState`. (Technically, having either `join` or `meet` should be sufficient since we cannot reuse the same lattice in two analyses, the lattice class is used as identifier of the analysis and `join`/`meet` are dual).

In practice, I see `setToEntry/ExitState` being used _in the framework_ for unanalyzable control flow situations: when we don't know all callsites of a function (e.g. the function is public and may be called from another translation unit), or when we don't know block or region predecessors/successors for forward/backward analysis respectively. These are the kind of irrecoverable situations where we would want to set the lattice instances to some edge state. I'm not sure which one... One interpretation is that in _both_ optimistic and pessimistic we could set the lattice to the $\top$ state, because it is always a conservatively correct overapproximation of the value. Optimistic analysis will be able to refine the approximation by approaching $\bottom$, but pessimistic analysis will be not be. Any thoughts on this? We are clearly missing one or two bits.

For initialization specifically, we may want to provide overrides for annotated functions and potentially mechanism to further customize other cases of control flow that are not analyzable through interfaces but may be specified by the user for specific analyses. That may be enough for activity/aliasing, but I'd really like to get to the bottom of the optimistic/pessimistic fixpoint story (pun intended).